File: cpp/examples/linux/rtc/DS3231/demo/main.cpp
#include <cstdio>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "DS3231.h"

// Backup-clock module for a data logger: reseeds the clock after a power
// loss, then logs a "reading" on every once-per-minute and once-per-hour
// alarm match, using the on-chip temperature sensor as the payload.

static volatile bool alarm1Fired = false, alarm2Fired = false;
static uint8_t lastStatus = 0;

static void onAlarm(uint8_t status) {
    lastStatus = status;
    if (status & DS3231Full::SOURCE_ALARM1) alarm1Fired = true;
    if (status & DS3231Full::SOURCE_ALARM2) alarm2Fired = true;
}

int main() {
    I2CConnectionLinux connection(1, DS3231Minimal::I2C_ADDRESS);
    DS3231Full rtc(connection);

    // --- Detect a lost power reference and reseed if needed ---
    // A fresh chip, or one whose backup coin cell died, reports the
    // Oscillator Stop Flag set — its clock/calendar registers cannot be
    // trusted until reseeded from a known-good reference.
    if (rtc.oscillatorStopped()) {
        DS3231Minimal::DateTime reference{2026, 1, 1, 4, 0, 0, 0};
        rtc.setDatetime(reference);  // also clears the Oscillator Stop Flag
        printf("oscillator was stopped - reseeded from reference timestamp\n");
    }

    // --- Arm a once-per-minute log tick and a once-per-hour summary tick ---
    // Both alarms repeat automatically (their registers are never rewritten),
    // so once armed, no further configuration is needed between matches.
    DS3231Full::Alarm2 hourlyTick{0, 0, 0, false, DS3231Full::ALARM2_MATCH_MINUTES};
    DS3231Full::Alarm1 perMinuteViaAlarm1{0, 0, 0, 0, false, DS3231Full::ALARM1_MATCH_SECONDS};
    rtc.setAlarm1(perMinuteViaAlarm1);   // fires once per minute at :00 seconds
    rtc.setAlarm2(hourlyTick);           // fires once per hour at :00 minutes

    rtc.onInterrupt(onAlarm);
    rtc.enableInterrupt(DS3231Full::SOURCE_ALARM1 | DS3231Full::SOURCE_ALARM2);

    // --- Log every match for a fixed number of minute-ticks, then clean up ---
    int logged = 0;
    while (logged < 5) {
        if (alarm1Fired || alarm2Fired) {
            DS3231Minimal::DateTime dt;
            rtc.getDatetime(dt);
            float tempC = rtc.readTemperature();
            if (alarm1Fired) {
                printf("[minute] %04u-%02u-%02u %02u:%02u:%02u  %.2f C  (status=0x%02X)\n",
                       dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second,
                       (double)tempC, lastStatus);
                logged++;
            }
            if (alarm2Fired) {
                printf("[hourly] %04u-%02u-%02u %02u:%02u:%02u  %.2f C  (status=0x%02X)\n",
                       dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second,
                       (double)tempC, lastStatus);
            }
            alarm1Fired = alarm2Fired = false;
        }
        usleep(200000);
    }

    rtc.disableInterrupt(DS3231Full::SOURCE_ALARM1 | DS3231Full::SOURCE_ALARM2);
    rtc.offInterrupt();
    return 0;
}

File: rust/examples/linux/rtc/ds3231/demo/src/main.rs
use embedded_hal::delay::DelayNs;
use linux_embedded_hal::{I2cdev, Delay};
use periph::chips::rtc::{Ds3231Full, DateTime, Alarm1Match, Alarm2Match, SOURCE_ALARM1, SOURCE_ALARM2};

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut rtc = Ds3231Full::new(dev, 0x68).expect("init DS3231"); // Create DS3231 Full driver, (i2c, addr=0x68)
    let mut delay = Delay;

    // --- Reseed the clock if the oscillator ever lost power ---
    // OSF is set on first power-up or any time the coin cell couldn't keep
    // the oscillator running. A fresh/recovered chip can't be trusted to
    // hold a meaningful time, so this data logger falls back to a fixed
    // reference timestamp rather than trusting garbage registers.
    if rtc.oscillator_stopped().expect("oscillator_stopped") {
        rtc.set_datetime(DateTime {           // Set the calendar clock, (datetime) → ()
            year: 2026, month: 1, day: 1, weekday: 4, hour: 0, minute: 0, second: 0,
        }).expect("set_datetime");            // also clears OSF, since the time is now known-good
    }

    // --- Arm a once-per-minute and a once-per-hour alarm ---
    // Alarm 1 fires every minute (00 seconds) to log a frequent sample;
    // Alarm 2 fires every hour (00 minutes) to mark an hourly boundary in
    // the log. Both repeat automatically — their registers are never
    // rewritten — so this is a fire-and-forget periodic logging setup.
    rtc.set_alarm1(0, 0, 0, 0, Alarm1Match::Seconds).expect("set_alarm1"); // seconds=0 -> fires once per minute
    rtc.set_alarm2(0, 0, 0, Alarm2Match::HoursMinutes).expect("set_alarm2");
    rtc.enable_interrupt(SOURCE_ALARM1 | SOURCE_ALARM2).expect("enable_interrupt");

    // --- Poll for alarm matches and log each one ---
    // Rust drivers expose only poll_interrupt() (no callback subscription —
    // polling is always caller-managed here), so the demo runs its own
    // loop, checking CONTROL_STATUS every second. Each fired alarm is
    // logged with the current date/time and the on-chip temperature — a
    // one-line "log entry" — for a fixed number of minute-alarms.
    let mut minute_alarms = 0;
    while minute_alarms < 5 {
        let status = rtc.poll_interrupt().expect("poll_interrupt");
        if status & SOURCE_ALARM1 != 0 {
            let dt = rtc.get_datetime().expect("get_datetime");
            let temp = rtc.read_temperature().expect("read_temperature");
            println!("[minute] {:04}-{:02}-{:02} {:02}:{:02}:{:02}  {:.2} C",
                     dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second, temp);
            minute_alarms += 1;
        }
        if status & SOURCE_ALARM2 != 0 {
            let dt = rtc.get_datetime().expect("get_datetime");
            println!("[hourly boundary] {:04}-{:02}-{:02} {:02}:{:02}:{:02}",
                     dt.year, dt.month, dt.day, dt.hour, dt.minute, dt.second);
        }
        delay.delay_ms(1000);
    }

    // --- Shut the alarms back down before exiting ---
    rtc.disable_interrupt(SOURCE_ALARM1 | SOURCE_ALARM2).expect("disable_interrupt");
}
